<?xml version="1.0"?>
<launch>
  <node name="$(arg robot_name)_proxy" pkg="proxy" type="ros_proxy" args="$(arg robot_name)">
	  <remap from="$(arg robot_name)/front_rgbd/image_raw/compressed" to="$(arg robot_name)/front/image_raw/compressed"/>
    <remap from="$(arg robot_name)/odom" to="$(arg robot_name)/nonexistent_input_odom"/>
    <remap from="$(arg robot_name)/cmd_vel" to="$(arg robot_name)/cmd_vel_osgar_intentionally_incorrect" />
  </node>
  <!-- NAVIGATION -->
  <node pkg="gmapping" type="slam_gmapping" name="gmapping_node">
    <param name="map_frame"  value="odom"/>
    <param name="base_frame" value="X300L/base_link"/>
    <param name="transform_publish_period"  value="0"/>
    <remap from="scan" to="X300L/front_scan"/>
  </node>

  <node pkg="move_base" type="move_base" respawn="false" name="move_base">
    <rosparam file="$(find robot)/params_x4/costmap_common_params.yaml" command="load" ns="global_costmap" />
    <rosparam file="$(find robot)/params_x4/costmap_common_params.yaml" command="load" ns="local_costmap" />
    <rosparam file="$(find robot)/params_x4/move_base_params.yaml" command="load" />
    <rosparam file="$(find robot)/params_x4/local_costmap_params.yaml" command="load" />
    <rosparam file="$(find robot)/params_x4/global_costmap_params.yaml" command="load" />
    <rosparam file="$(find robot)/params_x4/base_local_planner_params.yaml" command="load" />
    <!--remap from="map" to="/rtabmap/grid_map" /-->
    <!--remap from="map" to="/map" /-->
    <remap from="odom" to="$(arg robot_name)/odom_fused" />
    <remap from="cmd_vel" to="$(arg robot_name)/cmd_vel_osgar"/>
  </node>

  <node name="drone_height" pkg="robot" type="drone_height.py">
    <remap from="scan_down" to="$(arg robot_name)/bottom_scan" />
    <remap from="scan_up" to="$(arg robot_name)/top_scan" />
    <remap from="cmd_vel_drone" to="$(arg robot_name)/cmd_vel" />
    <remap from="cmd_vel" to="$(arg robot_name)/cmd_vel_osgar"/>
    <remap from="odom" to="$(arg robot_name)/odom_fused" />
  </node>

  <node pkg="explore_lite" type="explore" respawn="false" name="explore" output="screen">
    <param name="robot_base_frame" value="X300L/base_link"/>
    <param name="costmap_topic" value="/move_base/global_costmap/costmap"/>
    <param name="costmap_updates_topic" value="/move_base/global_costmap/costmap_updates"/>
    <param name="visualize" value="true"/>
    <param name="planner_frequency" value="1.0"/>
    <param name="progress_timeout" value="30.0"/>
    <param name="potential_scale" value="1e-3"/>
    <param name="orientation_scale" value="0.0"/>
    <param name="gain_scale" value="1.0"/>
    <param name="transform_tolerance" value="0.3"/>
    <param name="min_frontier_size" value="1.0"/>
  </node>

  <!-- ICP Odometry based on planar lidar -->
  <node pkg="rtabmap_ros" type="icp_odometry" name="icp_odometry">
    <param name="database_path" type="string" value=""/>
    <param name="subscribe_scan" type="bool" value="true"/>
    <param name="frame_id" type="string" value="$(arg robot_name)"/>
    <param name="odom_frame_id" type="string" value="odom"/>
    <param name="publish_tf" type="bool" value="false"/>
    <param name="publish_null_when_lost" type="bool" value="false"/>

    <param name="Reg/Strategy" type="string" value="1"/>
    <param name="OdomF2M/BundleAdjustment" type="string" value="0"/>
    <param name="Icp/MaxTranslation" type="string" value="0.3"/>
    <param name="Icp/MaxCorrespondenceDistance" type="string" value="0.15"/>
    <!-- Inputs -->
    <remap from="scan" to="/$(arg robot_name)/front_scan"/>
    <!-- Outputs -->
    <remap from="odom" to="/rtabmap/scan/odom" />
    <remap from="odom_info" to="/rtabmap/scan/odom_info"/>
    <remap from="odom_last_frame" to="/rtabmap/scan/odom_last_frame"/>
    <remap from="odom_local_map" to="/rtabmap/scan/odom_local_map"/>
    <remap from="odom_local_scan_map" to="/rtabmap/scan/odom_local_scan_map"/>
    <remap from="odom_filtered_input_scan" to="/rtabmap/scan/odom_filtered_input_scan"/>
    <remap from="scan_cloud" to="/rtabmap/scan/scan_cloud"/>
  </node>

  <node pkg="rtabmap_ros" type="rgbd_odometry" name="rgbd_odometry">
    <param name="database_path" type="string" value=""/>
    <param name="frame_id" type="string" value="$(arg robot_name)"/>
    <param name="odom_frame_id" type="string" value="odom"/>
    <param name="publish_tf" type="bool" value="false"/>
    <param name="publish_null_when_lost" type="bool" value="false"/>
    <!-- Inputs -->
    <remap from="rgb/image" to="$(arg robot_name)/front/optical/image_raw"/>
    <remap from="rgb/camera_info" to="$(arg robot_name)/front/optical/camera_info"/>
    <remap from="depth/image" to="$(arg robot_name)/front/optical/depth"/>
    <!-- Outputs -->
    <remap from="odom" to="/rtabmap/rgbd/odom" />
    <remap from="odom_info" to="/rtabmap/rgbd/odom_info"/>
    <remap from="odom_last_frame" to="/rtabmap/rgbd/odom_last_frame"/>
    <remap from="odom_local_map" to="/rtabmap/rgbd/odom_local_map"/>
    <remap from="odom_local_scan_map" to="/rtabmap/rgbd/odom_local_scan_map"/>
  </node>

  <node name="altimeter" pkg="robot" type="altitude_from_atmospheric_pressure.py" args="odom">
    <remap from="atmospheric_pressure" to="$(arg robot_name)/air_pressure"/>
    <remap from="pose" to="$(arg robot_name)/altitude"/>
  </node>

  <node name="pose_from_artifact_origin" pkg="proxy" type="pose_from_artifact_origin" args="$(arg robot_name) odom">
    <remap from="pose_from_origin" to="$(arg robot_name)/pose_from_origin"/>
  </node>

  <node name="imu_pruner" type="drop" pkg="topic_tools" args="$(arg robot_name)/imu/data 4 5 $(arg robot_name)/imu/pruned" />

  <!-- Fusion of ICP odometry, RGBD odometry and altitude with IMU data.
       https://github.com/introlab/rtabmap_ros/blob/master/launch/tests/sensor_fusion.launch
       https://answers.ros.org/question/248109/integrating-an-imu-with-rtabmap/
       http://docs.ros.org/melodic/api/robot_localization/html/state_estimation_nodes.html
  -->
  <node pkg="robot_localization" type="ukf_localization_node" name="fused_odometry">
    <param name="frequency" value="50"/>
    <param name="sensor_timeout" value="0.1"/>
    <param name="two_d_mode" value="false"/>

    <param name="odom_frame" value="odom"/>
    <param name="base_link_frame" value="$(arg robot_name)"/>
    <param name="world_frame" value="odom"/>

    <param name="transform_time_offset" value="0.0"/>

    <!-- ICP odometry from 2d lidar -->
    <param name="odom0" value="/rtabmap/scan/odom"/>
    <!-- RGBD odometry from depth camera -->
    <param name="odom1" value="/rtabmap/rgbd/odom"/>
    <!-- Altitude estimated from air pressure -->
    <param name="pose0" value="$(arg robot_name)/altitude"/>
    <!-- Information about location in the staging area. -->
    <param name="pose1" value="$(arg robot_name)/pose_from_origin"/>
    <!-- IMU -->
    <param name="imu0" value="$(arg robot_name)/imu/pruned"/>

    <param name="odom0_differential" value="false"/>
    <param name="odom1_differential" value="false"/>
    <param name="pose0_differential" value="false"/>
    <param name="pose1_differential" value="false"/>
    <param name="imu0_differential" value="false"/>

    <param name="odom0_relative" value="false"/>
    <param name="odom1_relative" value="false"/>
    <param name="pose0_relative" value="true"/>
    <param name="pose1_relative" value="true"/>
    <param name="imu0_relative" value="false"/>

    <param name="imu0_remove_gravitational_acceleration" value="true"/>

    <!-- The order of the values is x, y, z, roll, pitch, yaw, vx, vy, vz, vroll, vpitch, vyaw, ax, ay, az. -->
    <!-- We barely trust the ICP odometry on 2d scan with X and Y coordinates. Totally not anything more 3d. -->
    <rosparam param="odom0_config">[false, false, false,
                                    false, false, false,
                                    true, true, false,
                                    false, false, false,
                                    false, false, false]</rosparam>
    <rosparam param="odom1_config">[false, false, false,
                                    false, false, false,
                                    true, true, true,
                                    false, false, false,
                                    false, false, false]</rosparam>
    <rosparam param="pose0_config">[false, false, true,
                                    false, false, false,
                                    false, false, false,
                                    false, false, false,
                                    false, false, false]</rosparam>
    <rosparam param="pose1_config">[true, true, true,
                                    true, true, true,
                                    false, false, false,
                                    false, false, false,
                                    false, false, false]</rosparam>
    <rosparam param="imu0_config">[false, false, false,
                                   true,  true,  true,
                                   false, false, false,
                                   true,  true,  true,
                                   true,  true,  true] </rosparam>

    <!-- When some data arrive late, we are willing to go back in time and correct the estimate. -->
    <param name="smooth_lagged_data" value="true"/>
    <param name="history_length" value="1.0"/>

    <!--
    <param name="print_diagnostics" value="true"/>
    <remap from="/diagnostics" to="$(arg robot_name)/odom_diagnostics"/>
    -->

    <remap from="odometry/filtered" to="/$(arg robot_name)/odom_fused"/>
  </node>
</launch>
